src/globus_cli/parsing/param_types/notify_param.py

```diff
@@ -0,0 +1,117 @@
+        - "failed", "succeeded", "inactive" are normal vals
+
+        The converted value is always a dictionary which is expected to be
+        '**'-expanded into arguments to TransferData or DeleteData, containing
+        0-or-more of the following keys:
+
+          - notify_on_failed
+          - notify_on_inactive
+          - notify_on_succeeded
+        """
+        # Build a set of all values, excluding empty strings.
+        values = {v.strip() for v in value.lower().split(",")} - {""}
+
+        # You'll get an empty set if value is "" or "," to start with.
+        # Special-case it into "off", which helps avoid surprising scripts
+        # which take notification settings as inputs and build '--notify ""'.
+        if not values:
+            values = {"off"}
+
+        # Disallow invalid values.
+        invalid_values = values - {"off", "on", "succeeded", "failed", "inactive"}
+        if invalid_values:
+            raise click.UsageError(
+                f"--notify received these invalid values: {list(invalid_values)}"
+            )
+
+        # Disallow combining "off" with other values.
+        values_other_than_off = values - {"off"}
+        if "off" in values and values_other_than_off:
+            raise click.UsageError('--notify cannot accept "off" and another value')
+
+        # return the notification options to send!
+        # on means don't set anything (default)
+        if "on" in values:
+            return {}
+        # off means turn off everything
+        if "off" in values:
+            return {
+                "notify_on_succeeded": False,
+                "notify_on_failed": False,
+                "notify_on_inactive": False,
+            }
+
+        # otherwise, return the exact set of values seen
+        return {
+            "notify_on_succeeded": "succeeded" in values,
+            "notify_on_failed": "failed" in values,
+            "notify_on_inactive": "inactive" in values,
+        }
+
+    def shell_complete(
+        self, ctx: click.Context, param: click.Parameter, incomplete: str
+    ) -> list[CompletionItem]:
+        all_compoundable_options = ["failed", "inactive", "succeeded"]
+        all_options = ["on", "off"] + all_compoundable_options
+
+        # if the caller used `--notify <TAB>`, show all options
+        # if the caller used `--notify o<TAB>`, show `on` and `off`
+        # the logic below assumes there were commas
+        if "," not in incomplete:
+            return [CompletionItem(o) for o in all_options if o.startswith(incomplete)]
+
+        # grab the last partial name from the list
+        # e.g. if the caller used `--notify inactive,succ<TAB>`, then
+        #      collect `succ` as the last incomplete fragment
+        #
+        # also collect the valid completed parts into a set for comparisons
+        last_incomplete_fragment = incomplete.split(",")[-1]
+        already_contains = set(incomplete.split(",")[:-1])
```

Comment:
For me, the left-hand-size unpacking syntax falls into the same category as `for-else`.
I can read it when I see it, but it has never made it into my own vocabulary for writing!

I don't know that I frequently see cases for it as good as this one. Let's apply this rewrite. I'll read the other comment to understand your thoughts regarding leading commas.

Reply:
It occurs to me that, rather than using `.lstrip()`, outright filtering might be a better play, specifically for cases like `"succeeded,,failing,,"`:

```python
*existing_options, fragment_option = incomplete.split(",")
existing_options = [i for i in existing_options if i]

possible_options = sorted(set(all_compoundable_options) - set(existing_options))

return [
    CompletionItem(",".join(existing_options + [o]))
    for o in possible_options
    if o.startswith(fragment_option)
]
```